### PR TITLE
feat(bcc): support non syscall function kprobe

### PR DESCRIPTION
**Pull request by oojimmy**

### CODE REVIEW

File: src/agent/ebpf/bcc/bcc.go [hunk 1]
```diff
@@ -83,12 +83,16 @@ func (m *module) attachKProbe(probe *ebpfpb.ProbeSpec) error {
 	syscallName := bcc.GetSyscallFnName(probe.Target)
 	if probe.Entry != "" {
 		if err := m.attachKEntryProbe(syscallName, probe.Entry); err != nil {
-			return err
+			if err := m.attachKEntryProbe(probe.Target, probe.Entry); err != nil {
```
Comment (nascentcore-eng) [inline src/agent/ebpf/bcc/bcc.go]: This new code should be the "more correct" logic, because the old code incorrectly treat kprobe as syscall_probe, which is imprecise. Kprobe includes syscall and other kernel function

Comment (nascentcore-eng) [inline src/agent/ebpf/bcc/bcc.go]: This new code should be the "more correct" logic, because the old code incorrectly treat kprobe as syscall_probe, which is imprecise. Kprobe includes syscall and other kernel function

Comment (nascentcore-eng) [reply src/agent/ebpf/bcc/bcc.go]: So a fix on top of this current code is:
Check if probe name starts with syscall__
If yes, apply getsyscallname on probe.target 
Otherwise just attach with the supplied Target

Comment (oojimmy) [reply src/agent/ebpf/bcc/bcc.go]: Using filed prefixes to distinguish the type looks a little magic, It will increase the use of the user cost. because this is the proprietary method in starship. But it seems we do not have any better solution. So as you wish.

Comment (oojimmy) [reply src/agent/ebpf/bcc/bcc.go]: I will update doc in same time.

File: src/agent/ebpf/bcc/bcc.go [hunk 2]
```diff
@@ -12,6 +13,8 @@ import (
 
 	"github.com/tricorder/src/agent/ebpf/common"
 
+	utils "github.com/tricorder/src/utils/common"
```
Comment (nascentcore-eng) [inline src/agent/ebpf/bcc/bcc.go]: commonutils

Might be more consistent with the package path, WDYT?

File: src/agent/ebpf/bcc/bcc.go [hunk 3]
```diff
@@ -80,14 +83,19 @@ func (m *module) attachKProbe(probe *ebpfpb.ProbeSpec) error {
 	if len(probe.Target) == 0 {
 		return fmt.Errorf("while attaching kprobe '%v', target cannot be empty", probe)
 	}
-	syscallName := bcc.GetSyscallFnName(probe.Target)
+
+	target := probe.Target
+	if strings.HasPrefix(target, "sys_") {
```
Comment (nascentcore-eng) [inline src/agent/ebpf/bcc/bcc.go]: Should be checking 'syscall__'

Also, remember to document this requirements in the doc of implementing modules. 
(IIRC, it's also assigned to you).

I.e., syscall probes must be named with syscall__ prefix

File: src/agent/ebpf/bcc/bcc.go [hunk 4]
```diff
@@ -80,14 +83,19 @@ func (m *module) attachKProbe(probe *ebpfpb.ProbeSpec) error {
...
+		target = bcc.GetSyscallFnName(utils.StrTrimPrefix(target, len("sys_")))
```
Comment (nascentcore-eng) [inline src/agent/ebpf/bcc/bcc.go]: Actually, you should be checking probe name, not probe target name.

Probe name refers to the name of the probe function that will be triggered by an invocation of the attached target (function), uprobe kprobe tracepoint etc.

File: src/agent/ebpf/bcc/program_test.go [hunk 5]
```diff
@@ -48,7 +48,7 @@ func TestLoadAndPollData(t *testing.T) {
 		Probes: []*ebpfpb.ProbeSpec{
 			{
 				Type:   ebpfpb.ProbeSpec_KPROBE,
-				Target: "read",
+				Target: "sys_read",
```
Comment (nascentcore-eng) [inline src/agent/ebpf/bcc/program_test.go]: As I commented above, we'd check syscall__ prefix in the probe function

Not the probe target

File: src/agent/ebpf/bcc/bcc.go [hunk 6]
```diff
@@ -80,14 +83,19 @@ func (m *module) attachKProbe(probe *ebpfpb.ProbeSpec) error {
 	if len(probe.Target) == 0 {
 		return fmt.Errorf("while attaching kprobe '%v', target cannot be empty", probe)
 	}
-	syscallName := bcc.GetSyscallFnName(probe.Target)
+
+	target := probe.Target
+	if strings.HasPrefix(target, "syscall_") {
```
Comment (nascentcore-eng) [inline src/agent/ebpf/bcc/bcc.go:88]: You should document in the doc for writing ebpf code, otherwise it's hard to follow.

My previous comment actually meant to use code like below:
```
if strings.HasPrefix(probe.Entry, "syscall__") {
  target = bcc.GetSYscallFnName(...)
}
```
See https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#8-system-call-tracepoints
BCC applies this special log for probe function names like `syscall__`

But reading this, I guess a more explicit approach is to define a new type `SYSCALL_PROBE` in https://github.com/tricorder-observability/starship/blob/4f70262907a1e2342bb54c90d4d15a2197cb48be/src/pb/module/ebpf/ebpf.proto#L12
So that we do not need to bother with the naming tricks.

WDYT?

**Review by nascentcore-eng — APPROVED** [review]

File: src/agent/ebpf/bcc/bcc.go (hunk 6 above)
Comment (oojimmy) [reply src/agent/ebpf/bcc/bcc.go:88]: I agree with adding the new proto type to do this, I will do this in the next PR.
